DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,308 @@
+
+Instead, consider structuring your code to have dynamically constructable behavior logic. This allows you to isolate booleans to construction logic, a distinct concern, and construct the behavior you require for your context. 
+
+An example of code designed to be dynamically constructed may include the [decorator](https://refactoring.guru/design-patterns/decorator) pattern. It may use a [builder](https://refactoring.guru/design-patterns/builder) or an [abstract factory](https://refactoring.guru/design-patterns/abstract-factory) with a builder to construct the logic using boolean flag configuration read from CLI flags. There are many more combinations and, generally, you want to consider your problem space and design a piece of code that makes sense for your problem first.
+
+### Concurrency
+
+Concurrency introduces harder to rationalize execution relative to serialized programs and is easy to get wrong. Concurrency doesn't necessarily speed up a program either. With these points in mind, prefer synchronous APIs instead leaving concurrency to the caller. Not only does this eliminate a whole host of problems, it ensures your APIs are predictable and easier to use.
+
+Dave Cheney has made some excellent [comments](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#concurrency) on concurrency in Go.
+
+#### Channels
+
+If you find the need for channels prefer accepting them as parameters. Accepting channels as a parameter allows consumers to specify channel properties, such as buffered or unbuffered, creating a more flexible API. 
+
+Clearly document channel ownership thereby documenting who is responsible for closing it. If there are multiple writers, no-one can close it therefore no-one owns it. Note channels don't _need_ to be closed, they will be garbage collected once the necessary scoping conditions are met.
+
+Clearly document channel behavior. For example, what does your algorithm do if it can't send data on a channel? Will it drop data, block until the channel is writable, or something else?
+
+Avoid using channels for data mutation operations where a traditional mutex would suffice. 
+
+###  Comments
+
+Functions and variables should be named so that their purpose is clear. However, good naming and well structured code isn't justification to not document APIs. All public package APIs should be documented including those that may seem trivial. This, at minimum, helps with code documentation generation. Additionally, each package should have a clear purpose that can be documented with examples using the [package documentation feature of Go](https://go.dev/doc/effective_go#commentary).
```

Comment:
Note: If everyone agrees with this, we should probably look into adding some linter to prevent new public entities/functions from being added without doc comments.

Reply:
Made a note.